build/ci/downloadRenderer.js

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+'use strict';
+
+const fs = require('fs-extra');
+const path = require('path');
+const tmp = require('tmp');
+const { ExtensionRootDir } = require('../constants');
+const download = require('download');
+const StreamZip = require('node-stream-zip');
+
+async function unzip(zipFile, targetFolder) {
+    await fs.ensureDir(targetFolder);
+    return new Promise((resolve, reject) => {
+        const zip = new StreamZip({
+            file: zipFile,
+            storeEntries: true
+        });
+        zip.on('ready', async () => {
+            zip.extract('extension', targetFolder, (err) => {
+                if (err) {
+                    reject(err);
+                } else {
+                    resolve();
+                }
+                zip.close();
+            });
+        });
+    });
+}
+
+async function downloadRendererExtension() {
+    console.log('Downloading Renderer extension...');
+    const extensionName = 'ms-notebook-renderers.vsix';
+    const uri = `https://pvsc.blob.core.windows.net/extension-builds-jupyter-public/ms-notebook-renderers.vsix`;
+    let cleanup;
+    try {
+        const downloadDir = await new Promise((resolve, reject) => {
+            tmp.dir((err, dir, clean) => {
+                cleanup = clean;
+                if (err) {
+                    return reject(err);
+                }
+                resolve(dir);
+            });
+        });
+
+        const downloadedFile = path.join(downloadDir, extensionName);
+        if (fs.existsSync(downloadedFile)) {
+            await fs.unlink(downloadedFile);
+        }
+        await download(uri, downloadDir, { filename: extensionName });
+        console.log('Downloaded Renderer extension');
+
+        console.log('Extracting Renderer extension');
+        const extensionDir = path.join(downloadDir, 'extensionDir');
+        await unzip(downloadedFile, extensionDir)
+        console.log('Extracted Renderer extension');
+
+        await fs.copy(path.join(extensionDir, 'out', 'client_renderer'), path.join(ExtensionRootDir, 'out', 'client_renderer'));
```

Comment:
The problem with this is that npm run clean will destroy the renderer. So if I clean my output the extension goes away.  Perhaps we can put this and the bcrypt thing (which I have the same problem with) in a different directory?

Reply:
Or fix npm run clean to not delete it.

Reply:
The bcrypt exe is downloaded to src and then copied to out on each compile for this reason. (It's not supposed to go missing. If it is that's a bug)

Reply:
@joyceerhl I usually do this:

- git checkout main
- git pull
- git checkout -b rchiodo/<new branch>
- npm ci (this copies bcrypt to the out folder)
- npm run clean (this deletes it)

Building the source doesn't seem to recopy it as far as I can tell.

Reply:
I would prefer to keep all output in the `out` folder.
I'll create a separate issue to address this for both renderer & bycrypt.

One option is to setup a VSC task (thats a depednency of the compile task) that'll check if this goes missing & then download them auto magically... (using nodemon, pretty standard in node to use nodemon for such things).

Reply:
> Building the source doesn't seem to recopy it as far as I can tell.

@rchiodo How do you build the source, what command do you use? Gulp, VSC Task, `npm run compile` or other.

Reply:
I use the build tasks in VS code.